cudax/include/cuda/experimental/__async/sender/run_loop.cuh

```diff
@@ -187,93 +185,53 @@ public:
+    [[nodiscard]] _CUDAX_API auto schedule() const noexcept -> __sndr_t
     {
-      return __schedule_task{__loop_};
+      return __sndr_t{__loop_};
     }
 
-    _CUDAX_API friend bool operator==(const __scheduler& __a, const __scheduler& __b) noexcept
+    [[nodiscard]] _CUDAX_API friend bool operator==(const scheduler& __a, const scheduler& __b) noexcept
     {
       return __a.__loop_ == __b.__loop_;
     }
 
-    _CUDAX_API friend bool operator!=(const __scheduler& __a, const __scheduler& __b) noexcept
+    [[nodiscard]] _CUDAX_API friend bool operator!=(const scheduler& __a, const scheduler& __b) noexcept
     {
       return __a.__loop_ != __b.__loop_;
     }
   };
 
-  _CUDAX_API auto get_scheduler() noexcept -> __scheduler
+  [[nodiscard]] _CUDAX_API auto get_scheduler() noexcept -> scheduler
   {
-    return __scheduler{this};
+    return scheduler{this};
   }
 
-  _CUDAX_API void run();
-
-  _CUDAX_API void finish();
-
 private:
-  _CUDAX_API void __push_back(__task* __tsk);
-  _CUDAX_API auto __pop_front() -> __task*;
-
-  ::std::mutex __mutex{};
-  ::std::condition_variable __cv{};
-  __task __head{};
-  bool __stop = false;
-};
-
-template <class _Rcvr>
-_CUDAX_API inline void __operation<_Rcvr>::start() & noexcept {
-  _CUDAX_TRY(                                                               //
-    ({                                                                      //
-      __loop_->__push_back(this);                                           //
-    }),                                                                     //
-    _CUDAX_CATCH(...)                                                       //
-    ({                                                                      //
-      set_error(static_cast<_Rcvr&&>(__rcvr_), ::std::current_exception()); //
-    })                                                                      //
-    )                                                                       //
-}
-
-_CUDAX_API inline void run_loop::run()
-{
-  for (__task* __tsk = __pop_front(); __tsk != &__head; __tsk = __pop_front())
+  // Returns true if any tasks were executed.
+  _CUDAX_API bool __execute_all(bool __wait) noexcept
   {
-    __tsk->__execute();
+    // Wait until the queue has tasks to execute and then dequeue all of them.
+    auto __queue = __queue_.pop_all(__wait);
+    if (__queue.empty())
+    {
+      return false;
+    }
+    // Execute all the tasks in the queue.
+    for (auto __task : __queue)
+    {
+      __task->__execute();
+    }
+    __queue.clear();
+    return true;
   }
-}
 
-_CUDAX_API inline void run_loop::finish()
-{
-  ::std::unique_lock __lock{__mutex};
-  __stop = true;
-  __cv.notify_all();
-}
-
-_CUDAX_API inline void run_loop::__push_back(__task* __tsk)
-{
-  ::std::unique_lock __lock{__mutex};
-  __tsk->__next_ = &__head;
-  __head.__tail_ = __head.__tail_->__next_ = __tsk;
-  __cv.notify_one();
-}
+  _CUDA_VSTD::atomic<bool> __finishing_{false};
```

Comment:
I dont know atomics, but would `cuda::std::atomic_flag` be more appropriate?

Reply:
in `run_loop::run` we want to test whether the flag is set or not. `std::atomic_flag` only has `test_and_set` in C++17. a simple `test` method wasn't added until C++20.

Reply:
But `cuda::std::atomic_flag` should have all the jazz already in 17